log4j-to-jul/src/main/java/org/apache/logging/log4j/tojul/JULLogger.java

```diff
@@ -0,0 +1,246 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache license, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the license for the specific language governing permissions and
+ * limitations under the license.
+ */
+package org.apache.logging.log4j.tojul;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.logging.LogRecord;
+import java.util.logging.Logger;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.Marker;
+import org.apache.logging.log4j.message.Message;
+import org.apache.logging.log4j.message.MessageFactory;
+import org.apache.logging.log4j.spi.AbstractLogger;
+
+/**
+ * Implementation of {@link org.apache.logging.log4j.Logger} that's backed by a {@link Logger}.
+ *
+ * This implementation currently ignores {@link Marker}.
+ *
+ * @author <a href="http://www.vorburger.ch">Michael Vorburger.ch</a> for Google
+ */
+public class JULLogger extends AbstractLogger {
+    private static final long serialVersionUID = 1L;
+
+    // @VisibleForTesting
+    final Logger logger;
+
+    // This implementation is inspired by org.apache.logging.slf4j.SLF4JLogger
+
+    public JULLogger(final String name, final MessageFactory messageFactory, final Logger logger) {
+        super(name, messageFactory);
+        this.logger = requireNonNull(logger, "logger");
+    }
+
+    public JULLogger(final String name, final Logger logger) {
+        super(name);
+        this.logger = requireNonNull(logger, "logger");
+    }
+
+    @Override
+    public void logMessage(final String fqcn, final Level level, final Marker marker, final Message message, final Throwable t) {
+        java.util.logging.Level julLevel = convertLevel(level);
+        if (!logger.isLoggable(julLevel)) {
+            return;
+        }
+        LogRecord record = new LogRecord(julLevel, message.getFormattedMessage()); // TODO getFormat() or getFormattedMessage() ?
+        record.setLoggerName(getName()); // TODO getName() or fqcn parameter? What's that?
+        record.setParameters(message.getParameters());
```

Comment:
Good questions!

### 1. getFormattedMessage/getMessage/getParameters

JUL `LogRecord` and the log4j-api `Message` are similar in that they both provide a raw format string, and a parameter array. However, the log4j-api encodes how the format and parameters are combined in the `Message` implementation (`getFormattedMessage` + `formatTo(StringBuilder)` handle formatting) where JUL (default formatters anyhow) use [java.text.MessageFormat.format](https://github.com/openjdk/jdk/blob/dfacda488bfbe2e11e8d607a6d08527710286982/src/java.logging/share/classes/java/util/logging/Formatter.java#L92-L150).

Given that, we don't want to use `getFormat()` and definitely don't want to set the parameters here, otherwise there's JUL may do substitution on the user-provided message...and we don't want to go there again ;-)

### 2. Logger name/fqcn

We definitely want to set the logger name to `getName` which is the name of the logger as requested from the LogManager.
The `fqcn` is the fully qualified name of the logger implementation, e.g. `org.apache.logging.log4j.tojul.JULLogger`. This is useful to find the call-site, where we can traverse just above that fqcn. JUL supports this sort of thing using `getSourceClassName ` and `getSourceMethodName`, so it may be a good idea to support the functionality here, otherwise the `LogRecord` will always report `JULLogger.logMessage` as the caller. I'd subclass LogRecord to lazily evaluate the call-site when it's requested given the compute cost of StackWalker/etc.

Reply:
1. So KEEP `new LogRecord(julLevel, message.getFormattedMessage())` as-is, and REMOVE the `// TODO getFormat() or getFormattedMessage() ?` AND also remove the `record.setParameters(message.getParameters());` - am I getting this right? I've done this now, could you just double check I got this right and ACK 1. is fine.

2. I see. Let me get back to that point. (As I'm adding more test coverage I'm finding that other -more basic- things need a bit more work, first.)

Reply:
1: That's exactly right, thanks!

Reply:
2. Is now all that's really left to do here (from my side); I'll look more into on one of the next few days, and then un-draft this.

Reply:
Done. Resolving this thread, and undrafting - this PR is now ready for review, from my side.